test/src/test/java/hudson/model/AbstractBuildTest2.java

```diff
@@ -0,0 +1,85 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.model;
+
+import hudson.Launcher;
+import hudson.model.queue.QueueTaskFuture;
+import java.io.IOException;
+import java.util.Arrays;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.TestExtension;
+
+import static org.hamcrest.Matchers.*;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Unit tests of {@link AbstractBuild}.
+ * @author Oleg Nenashev
+ */
+public class AbstractBuildTest2 {
+    
+    @Rule
+    public JenkinsRule rule = new JenkinsRule();
+    
+    @Test
+    @Issue("JENKINS-30730")
+    @SuppressWarnings("deprecation")
+    public void reportErrorShouldNotFailForNionPublisherClass() throws Exception {
+        FreeStyleProject prj = rule.createFreeStyleProject();
+        prj.addProperty(new ErrorneousJobProperty());
+        QueueTaskFuture<FreeStyleBuild> future = prj.scheduleBuild2(0);     
+        assertThat("Build should be actually scheduled by Jenkins", future, notNullValue());
+        FreeStyleBuild build = future.get();
+        assertThat("Build should contain the error message", build.getLog(), 
+                stringContainsInOrder(Arrays.asList(ErrorneousJobProperty.ERROR_MESSAGE)));
+        assertThat("Build log should not contain the class cast error", build.getLog(), 
+                not(stringContainsInOrder(Arrays.asList(ClassCastException.class.getName()))));
+    }
+    
+    /**
+     * Job property, which always fails with an exception.
+     */
+    public static class ErrorneousJobProperty extends JobProperty<FreeStyleProject> {
+
+        public static final String ERROR_MESSAGE = "This publisher fails by design";
+        
+        @Override
+        public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+            throw new IOException(ERROR_MESSAGE);
+        }
+        
+        @TestExtension("reportErrorShouldNotFailForNionPublisherClass")
+        public static class DescriptorImpl extends JobPropertyDescriptor {
+
+            @Override
+            public String getDisplayName() {
+                return "Always throws exception in perform()";
```

Comment:
I wonder if we should not simply make `Descriptor.getDisplayName` default to `return clazz.getSimpleName();` so tests do not have to bother with this boilerplate. (Though in case I think you actually wanted to verify that the log contains this `displayName`, right?)